src/lib/hooks/swap/useBestTrade.ts

```diff
@@ -0,0 +1,76 @@
+import { Currency, CurrencyAmount, TradeType } from '@uniswap/sdk-core'
+import { useClientSideV3Trade } from 'hooks/useClientSideV3Trade'
+import useDebounce from 'hooks/useDebounce'
+import { useMemo } from 'react'
+import { InterfaceTrade, TradeState } from 'state/routing/types'
+
+import useClientSideSmartOrderRouterTrade from '../routing/useClientSideSmartOrderRouterTrade'
+
+/**
+ * Returns the best v2+v3 trade for a desired swap.
+ * @param tradeType whether the swap is an exact in/out
+ * @param amountSpecified the exact amount to swap in/out
+ * @param otherCurrency the desired output/payment currency
+ */
+export function useBestTrade(
+  tradeType: TradeType,
+  amountSpecified?: CurrencyAmount<Currency>,
+  otherCurrency?: Currency
+): {
+  state: TradeState
+  trade: InterfaceTrade<Currency, Currency, TradeType> | undefined
+} {
+  // debounce used to prevent excessive requests to SOR, as it is data intensive
+  // this helps provide a "syncing" state the UI can reference for loading animations
+  const [debouncedAmount, debouncedOtherCurrency] = useDebounce(
+    useMemo(() => [amountSpecified, otherCurrency], [amountSpecified, otherCurrency]),
+    200
+  )
+
+  const routingAPITrade = useClientSideSmartOrderRouterTrade(tradeType, debouncedAmount, debouncedOtherCurrency)
+
+  // check if the artifical delay we impose is still occuring
+  const debouncing = amountSpecified !== undefined && debouncedAmount === undefined
+
+  // the amount associated with field user is typing in
+  const amountFromLatestTrade =
+    tradeType === TradeType.EXACT_INPUT ? routingAPITrade.trade?.inputAmount : routingAPITrade.trade?.outputAmount
+
+  // ordered currencies from API trade
+  const [currencyFromTrade, otherCurrencyFromTrade] =
```

Comment:
there's no reason you can't destructure all 3 together (currencies and amountFromLatestTrade). I think it'd be clearer if you did, but did so as part of an if/else statement. You can do so as a separate helper function so that you are returning in each case and don't need to use `let`.

```
const [amount, currency, otherCurrency] = getTradeInputs(routingApiTrade.trade)
```